### PR TITLE
Building system

### PR DESCRIPTION
**Pull request by andreafra**
Added:
- Function to place blocks, complete with check for duplicate blocks
- Debug Utility functions to test quickly new stuff
- A data structure (class ```Block```) to keep track of placed blocks
Knows Issues:
- Unity spawns two Grid objects, dunno why

### CODE REVIEW

File: super-cool-game/Assets/Scripts/Building/BuildingController.cs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        // TODO: Find why unity spawns two gameobjects, Block Grid and Block Grid (clone).
+        grid = new GameObject();
+        grid.name = "Block Grid";
+        grid = Instantiate(grid, mapCenter, Quaternion.identity);
+    }
+
+    // <summary>
+    // Create a singleton to call functions without instancing the class directly
+    // </summary>
+    private static BuildingController _instance = null;
```
Comment (aleasto) [inline super-cool-game/Assets/Scripts/Building/BuildingController.cs]: This is ugly and unnecessary. You can easily reference the script with GetComponent on the parent gameobject

File: super-cool-game/Assets/Scripts/Building/BuildingController.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        // If there's no block in that position, create a new one.
+        if (!placedBlocks.Exists(
+            x => x.position == Block.GetBlockRelativeCoordinates(position)
+            )) {
+            // Spawn the block
+            GameObject go = Instantiate(blocks[type], position, Quaternion.identity);
+            go.transform.parent = grid.transform;
+
+
+
```
Comment (aleasto) [inline super-cool-game/Assets/Scripts/Building/BuildingController.cs:85]: so much white space

File: super-cool-game/Assets/Scripts/Building/BuildingController.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        // Calculate center of the map
+        mapCenter = new Vector2(ground.transform.position.x, ground.transform.position.y + (groundSize.y / 2));
+
+        // Spawn two empty gameObjects.
+        // Blocks must be children of these two.
+
+        // TODO: Find why unity spawns two gameobjects, Block Grid and Block Grid (clone).
+        grid = new GameObject();
+        grid.name = "Block Grid";
+        grid = Instantiate(grid, mapCenter, Quaternion.identity);
```
Comment (aleasto) [inline super-cool-game/Assets/Scripts/Building/BuildingController.cs]: new GameObject() already instantiates it. This creates a duplicate

File: super-cool-game/Assets/Scripts/Player/PlayerMovement.cs [hunk 4]
```diff
@@ -31,7 +31,7 @@ public class PlayerMovement : MonoBehaviour {
 
     void ComputeMovement () {
         // Resultant of movement
-        rb.velocity = new Vector2(moveX, rb.velocity.y);
+        // rb.velocity = new Vector2(moveX * speed, rb.velocity.y);
```
Comment (aleasto) [inline super-cool-game/Assets/Scripts/Player/PlayerMovement.cs]: Don't touch this here

File: super-cool-game/Assets/Scripts/Building/Block.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    // </summary>
+    public static Vector2Int GetBlockRelativeCoordinates(Vector2 position) {
+        int blockW = BuildingController.blockWidth;
+        // Blocks in the grid are in this form
+        // ...[-2][-1](center)[1][2]...
+        int x = (int)(Mathf.Sign(position.x) * Mathf.Ceil(Mathf.Abs(position.x) / blockW));
+        Debug.Log(x);
+        return new Vector2Int(
+            (int)(Mathf.Sign(position.x) * Mathf.Ceil(Mathf.Abs(position.x) / blockW)),
+            (int)(Mathf.Sign(position.y) * Mathf.Floor(Mathf.Abs(position.y) / blockW)));
```
Comment (aleasto) [inline super-cool-game/Assets/Scripts/Building/Block.cs:43]: just cast position.x/blockW to an int

File: super-cool-game/Assets/Scripts/Building/BuildingController.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+// Each kind of block has a numerical id
+// that is used to keep track of disposition
+// of the block in the grid that is shared between
+// client and server.
+// <summary>
+// Return a numerical value for each kind of block.
+// </summary>
+enum BlockType {
+    Spawn = 0,
+    Heal = 1
```
Comment (aleasto) [inline super-cool-game/Assets/Scripts/Building/BuildingController.cs]: Counting from 0 is default on enums, remove assignments.

File: super-cool-game/Assets/Scripts/Building/BuildingController.cs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    // Reference grid of block gameobjects
+    // this is DATA
+    [HideInInspector] public List<Block> placedBlocks = new List<Block>();
+
+    // Container for blocks in world space, in order to have relative positioning
+    // this is GAMEOBJECTS
+    [HideInInspector] public GameObject grid;
+
+    void Start() {
+        Vector2 groundSize = ground.transform.GetComponent<SpriteRenderer>().bounds.size;
```
Comment (aleasto) [inline super-cool-game/Assets/Scripts/Building/BuildingController.cs]: Remove .transform.

File: super-cool-game/Assets/Scripts/Building/BuildingController.cs [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    // <summary>
+    // This function place a block of the specified type in the scene,
+    // at the coordinates of the Vector2 that are passed to it.
+    // </summary>
+    // It should be the server to decide where and when place the block.
+    public void PlaceBlock(int type, Vector2 position) {
+
+        // TODO: Check if there isnt already a block there.
```
Comment (aleasto) [inline super-cool-game/Assets/Scripts/Building/BuildingController.cs]: Isn't it done?

File: super-cool-game/Assets/Scripts/Building/BuildingController.cs [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        // If there's no block in that position, create a new one.
+        if (!placedBlocks.Exists(
+            x => x.position == Block.GetBlockRelativeCoordinates(position)
+            )) {
+            // Spawn the block
+            GameObject go = Instantiate(blocks[type], position, Quaternion.identity);
+            go.transform.parent = grid.transform;
```
Comment (aleasto) [inline super-cool-game/Assets/Scripts/Building/BuildingController.cs]: Remove .transform. and .transform
Also maybe rename _go_ to something meaningful

File: super-cool-game/Assets/Scripts/Player/PlayerBuilding.cs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    // The rounded position to the size of the grid which is multiples of blockWidth
+    public Vector2 SnappedPosition {
+        get {
+            return new Vector2(
+                Mathf.Ceil(transform.position.x / blockWidth) * blockWidth - blockWidth / 2,
+                Mathf.Ceil(transform.position.y / blockWidth) * blockWidth - blockWidth / 2);
+        }
+    }
+
+    private int blockWidth = BuildingController.blockWidth;
```
Comment (aleasto) [inline super-cool-game/Assets/Scripts/Player/PlayerBuilding.cs]: Move this up

File: super-cool-game/Assets/Scripts/Player/PlayerStats.cs [hunk 11]
```diff
@@ -0,0 +1,29 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class PlayerStats : MonoBehaviour {
+
+    public const int maxHealth = 100;
+    public int currentHealth;
```
Comment (aleasto) [inline super-cool-game/Assets/Scripts/Player/PlayerStats.cs]: Bad idea to put this public.
If you need it public to be read, do `public int currentHealth { get; private set; }`
It should only be writable through appropriate methods

File: super-cool-game/Assets/Scripts/UI/DebugUIController.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    private void Start() {
+        pb = player.GetComponent<PlayerBuilding>();
+    }
+    public void PlaceBlock() {
+        if (pb.isBuilding) {
+            BuildingController.instance.PlaceBlock((int)BlockType.Spawn, pb.SnappedPosition);
+        }
+    }
+
+    public void StartBuilding() {
```
Comment (aleasto) [inline super-cool-game/Assets/Scripts/UI/DebugUIController.cs]: ToggleBuilding would be more appropriate

**Review by aleasto — CHANGES_REQUESTED** [review]

File: super-cool-game/Assets/Scripts/Building/Block.cs (hunk 5 above)
Comment (aleasto) [reply super-cool-game/Assets/Scripts/Building/Block.cs:43]: Ah, you don't want 0, nevermind then

File: super-cool-game/Assets/Scripts/Building/BuildingController.cs (hunk 2 above)
Comment (andreafra) [reply super-cool-game/Assets/Scripts/Building/BuildingController.cs:85]: LMAO

File: super-cool-game/Assets/Scripts/Building/BuildingController.cs (hunk 9 above)
Comment (andreafra) [reply super-cool-game/Assets/Scripts/Building/BuildingController.cs]: [See here](https://docs.unity3d.com/ScriptReference/Transform-parent.html)

File: super-cool-game/Assets/Scripts/Building/Block.cs [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+    // <summary>
+    //  Return a Vector2Int that doesnt represent the actual block coordinates,
+    //  but its position in a grid where the cells have width and height of 1.
+    // </summary>
+    public static Vector2Int GetBlockRelativeCoordinates(Vector2 position) {
+        int blockW = BuildingController.blockWidth;
+        // Blocks in the grid are in this form
+        // ...[-2][-1](center)[1][2]...
+        int x = (int)(Mathf.Sign(position.x) * Mathf.Ceil(Mathf.Abs(position.x) / blockW));
```
Comment (aleasto) [inline super-cool-game/Assets/Scripts/Building/Block.cs]: x is not used

**Review by aleasto — CHANGES_REQUESTED** [review]

File: super-cool-game/Assets/Scripts/Building/Block.cs (hunk 13 above)
Comment (andreafra) [reply super-cool-game/Assets/Scripts/Building/Block.cs]: OH NO

**Pull request closed without merging**